### ISSUE TITLE
Feature request: method for human readable units

### ISSUE THREAD

**Issue by ryanduve** [astropy/astropy#6913]
Would like to automatically determine the units for a quantity so it's more human readable.  For example, Googling "100000000000 bytes" (https://www.google.com/search?q=100000000000%20bytes) returns 100 GB automatically.

Currently, this functionality only exists if the desired units are specified, i.e., `(100000000000 * u.m).to(u.Gm)`, but it would be convenient to have a function that automatically converted `(100000000000 m).human_readable()` to 100 Gm.

**Comment by mhvk** [comment]
We do have something like this for bytes `astropy.utils.console.human_file_size`, but it works well only for single units; for compound ones like `kg m/s`, it becomes somewhat arbitrary which unit you give the prefix.

Also, do you want this specifically for printing? I'm wondering if it makes most sense in the form of an option for formatting, or even a context manager that automatically applies it to `__repr__`. I think it would be good to be sure we have a clear idea for an API, in case someone wants to actually implement it.

**Comment by ryanduve** [comment]
The use-case I had in mind was for graphing.  For example, if an ML model spits out 10 results that range from 5000 to 10000 seconds, I would want to change them to hours.  However, if a hyperparameter or something changes and the model spits out values ranging from 100000 to 500000 seconds, I would want to change units to days.


Currently, I do the obvious thing and just figure out the best unit and divide values accordingly, but specifying automatic "human readability" would be a major addition to the units component of astropy.

**Comment by mhvk** [comment]
OK, so it would be unit-specific, with time changing to minutes, hours, days, etc., as appropriate, but length being some prefixed meters.

It might be an idea to start with just defining a separate function rather than a method, so we have something to play with (say, `u.human_readable(...)`)

**Comment by ananyashreyjain** [comment]
@mhvk  isn't  `astropy.utils.console.human_time(..)`  already there to convert time in easily readable units?

**Comment by mhvk** [comment]
You're right -- so at least for bytes and time, we already have it; maybe the remainder could use a "default" that just picks the right SI prefix.

**Comment by mhvk** [comment]
Just noting that I got a separate request per e-mail for this functionality; looking back at #7033, that basically seemed close to what is needed, and it might be worth reviving it.

**Comment by mhvk** [comment]
After seeing the text in a duplicate issue, #12951, maybe good to summarize the state. I think there are two main problem that make it difficult to have a general solution for "human-readable" units.
1. What to do with composite units like `km/s` (does one change 'km' to 'Mm' or 's' to 'ms'?).
2. What to do with array-valued quantities.

Perhaps it should just work with simple prefix units only? And for arrays find some "optimum" from the mean or so?

Though since most of the use is in string output of units (see #12951), perhaps the most logical is to adjust the `to_string` method, perhaps following #7033 and https://github.com/gammapy/gammapy/pull/3752. One could then have keyword arguments to determine what exactly should be done. Here, we do need to make sure that whatever we add is consistent with what is currently being done by subclasses, such as `Angle.to_string()` (indeed, that might inspire some nice solution).  Obviously, another solution is just to start with a helper function.

**Comment by adonath** [comment]
> 1. What to do with composite units like km/s

I agree this is tricky. I think it would require to manually assign priorities / weights to prefixes, effectively making it an optimization problem. But this might be highly specific on the domain as well. In any case supporting this for `PrefixUnit` only would already be useful.

> 2. What to do with array-valued quantities

If all numbers are in the same range the mean should work fine. If the numbers vary by multiple orders of magnitudes it might work as well, but not necessarily. Alternatively one could do a "democratic decision" and take the preferred unit that works for the "majority" of entries in the array.

**Comment by eerovaher** [comment]
> What to do with array-valued quantities.

This code in the `Distance` constructor uses the mean of the array:
https://github.com/astropy/astropy/blob/da206faba923db66e53aac29450dd57cea211cc0/astropy/coordinates/distances.py#L132-L143

**Comment by mbsck** [comment]
I might be a bit late (and may be missing the problem here), but for a general function that does not cover any special cases like astronomical distances, wouldn't something like this work:
```python
def get_prefix(value: u.Quantity) -> str:
    """Gets the correct prefix for the unit."""
    base = value.decompose().value
    if base < 1e-27:
        return "q"
    if base < 1e-24:
        return "r"
    if base < 1e-21:
        return "y"
    if base < 1e-18:
        return "z"
    if base < 1e-15:
        return "a"
    if base < 1e-12:
        return "f"
    if base < 1e-9:
        return "p"
    if base < 1e-6:
        return "n"
    if base < 1e-3:
        return "u"
    if base < 1e-2:
        return "m"
    if base < 1e-1:
        return "c"
    if base < 1e1:
        return "d"
    if base < 1e2:
        return "da"
    if base < 1e3:
        return "h"
    if base < 1e6:
        return "k"
    if base < 1e9:
        return "M"
    if base < 1e12:
        return "G"
    if base < 1e15:
        return "T"
    if base < 1e18:
        return "P"
    if base < 1e21:
        return "E"
    if base < 1e24:
        return "Z"
    if base < 1e27:
        return "Y"
    if base < 1e30:
        return "R"
    return "Q"
```
Then with the prefix one could apply this to *one unit (at a time) only* in the following
```python
>>> x = 1e-3 * u.m
>>> prefix = get_prefix(x)
... "m"
>>> new_unit = Unit(f"{prefix}{x.unit}")
... Unit("mm")
>>> x_new = x.to(new_unit)
... <Quantity 1. mm>
```
and even to arrays with ranges not too far apart (~same scale)
```python
>>> array = [1e-6, 8e-6, 1.5e-5]
>>> array.to(f"{get_prefix(array.min()}{x.unit}")
... <Quantity [1, 8, 15] * um>
```
---
If there is a list of prefixes (or one could be implemented in `astropy`, then the function could become even shorter)
```python3
def get_prefix(value: u.Quantity, prefixes: List[str]) -> str:
    """Gets the correct prefix for the unit."""
    base = value.decompose().value
    for prefix, exp in  zip(prefixes, range(-27, 33, 3)):
        if base < 10**exp:
            return prefix
    return prefixes[-1]
```
---
This was inspired by [Astropy units: automatic conversion to the closest human readable unit](https://community.openastronomy.org/t/astropy-units-automatic-conversion-to-the-closest-human-readable-unit/285)
